Performance improvements on Trie handling

diega opened the pull request:
A refactor for splitting the Trie and its underlying store is planned but not tackled by this PR. Having that in mind, I don't think integrating this changes should make that task harder

rskops commented:
SonarQube analysis reported 24 issues
* ![CRITICAL][CRITICAL] 12 critical
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 8 minor
* ![INFO][INFO] 1 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L62](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L62): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L91](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L91): A "HashSet<byte[]>" cannot contain a "byte[]" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2175)
1. ![CRITICAL][CRITICAL] [HashMapDB.java#L105](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/HashMapDB.java#L105): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [HashMapDB.java#L124](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/HashMapDB.java#L124): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [HashMapDB.java#L125](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/HashMapDB.java#L125): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [HashMapDB.java#L127](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/HashMapDB.java#L127): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L186](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L186): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L195](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L195): Change this condition so that it does not always evaluate to "false" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2583)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L200](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L200): Change this condition so that it does not always evaluate to "false" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2583)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L263](https://github.com/rsksmart/rskj/blob/b980b164e2164ccb825136183ea5035c93cdfcf2/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L263): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

rskops commented:
SonarQube analysis reported 16 issues
* ![CRITICAL][CRITICAL] 7 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 1 info

#### Top 10 issues

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L93](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L93): A "Set<byte[]>" cannot contain a "byte[]" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2175)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L186](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L186): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L195](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L195): Change this condition so that it does not always evaluate to "false" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2583)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L200](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L200): Change this condition so that it does not always evaluate to "false" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2583)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L263](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L263): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L264](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L264): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L266](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L266): Missing curly brace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00121)
1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L69](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L69): Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1066)
1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L91](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L91): Method org.ethereum.datasource.DataSourceWithCache.addKeys(Set, HashMapDB) uses array as basis of collection [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AABC_ARRAY_BASED_COLLECTIONS)
1. ![MAJOR][MAJOR] [HashMapDB.java#L155](https://github.com/rsksmart/rskj/blob/c4e3342ef113281072f5ce50426ec12d260e9116/rskj-core/src/main/java/org/ethereum/datasource/HashMapDB.java#L155): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

rskops commented:
SonarQube analysis reported 8 issues

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L91](https://github.com/rsksmart/rskj/blob/3879f61deb7dea73ab00677393e1a03ef50b16c1/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L91): A "Set<byte[]>" cannot contain a "byte[]" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2175)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L198](https://github.com/rsksmart/rskj/blob/3879f61deb7dea73ab00677393e1a03ef50b16c1/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L198): Change this condition so that it does not always evaluate to "false" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2583)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L203](https://github.com/rsksmart/rskj/blob/3879f61deb7dea73ab00677393e1a03ef50b16c1/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L203): Change this condition so that it does not always evaluate to "false" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2583)
1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L89](https://github.com/rsksmart/rskj/blob/3879f61deb7dea73ab00677393e1a03ef50b16c1/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L89): Method org.ethereum.datasource.DataSourceWithCache.addKeys(Set, HashMapDB) uses array as basis of collection [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AABC_ARRAY_BASED_COLLECTIONS)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L337](https://github.com/rsksmart/rskj/blob/3879f61deb7dea73ab00677393e1a03ef50b16c1/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L337): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MINOR][MINOR] [DataSourceWithCache.java#L87](https://github.com/rsksmart/rskj/blob/3879f61deb7dea73ab00677393e1a03ef50b16c1/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L87): 2nd parameter 'map' could be declared as org.ethereum.datasource.KeyValueDataSource instead [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AOCP_OVERLY_CONCRETE_PARAMETER)
1. ![MINOR][MINOR] [HashMapDB.java#L133](https://github.com/rsksmart/rskj/blob/3879f61deb7dea73ab00677393e1a03ef50b16c1/rskj-core/src/main/java/org/ethereum/datasource/HashMapDB.java#L133): This block of commented-out lines of code should be removed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![INFO][INFO] [LevelDbDataSource.java#L198](https://github.com/rsksmart/rskj/blob/3879f61deb7dea73ab00677393e1a03ef50b16c1/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L198): Redundant nullcheck of value, which is known to be non-null in org.ethereum.datasource.LevelDbDataSource.put(byte[], byte[]) [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3ARCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

rskops commented:
SonarQube analysis reported 6 issues

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L91](https://github.com/rsksmart/rskj/blob/2d5a7c6abefd9331ec06414f5f654966e6291323/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L91): A "Set<byte[]>" cannot contain a "byte[]" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2175)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L198](https://github.com/rsksmart/rskj/blob/2d5a7c6abefd9331ec06414f5f654966e6291323/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L198): Method org.ethereum.datasource.LevelDbDataSource.put(byte[], byte[]) passes a concatenated string to SLF4J's format string [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALO_APPENDED_STRING_IN_FORMAT_STRING)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L203](https://github.com/rsksmart/rskj/blob/2d5a7c6abefd9331ec06414f5f654966e6291323/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L203): Method org.ethereum.datasource.LevelDbDataSource.put(byte[], byte[]) passes a concatenated string to SLF4J's format string [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALO_APPENDED_STRING_IN_FORMAT_STRING)
1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L89](https://github.com/rsksmart/rskj/blob/2d5a7c6abefd9331ec06414f5f654966e6291323/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L89): Method org.ethereum.datasource.DataSourceWithCache.addKeys(Set, HashMapDB) uses array as basis of collection [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AABC_ARRAY_BASED_COLLECTIONS)
1. ![MAJOR][MAJOR] [LevelDbDataSource.java#L337](https://github.com/rsksmart/rskj/blob/2d5a7c6abefd9331ec06414f5f654966e6291323/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L337): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1186)
1. ![MINOR][MINOR] [DataSourceWithCache.java#L87](https://github.com/rsksmart/rskj/blob/2d5a7c6abefd9331ec06414f5f654966e6291323/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L87): 2nd parameter 'map' could be declared as org.ethereum.datasource.KeyValueDataSource instead [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AOCP_OVERLY_CONCRETE_PARAMETER)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

rskops commented:
SonarQube analysis reported 4 issues

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L91](https://github.com/rsksmart/rskj/blob/a3f72ed52faf4ecc3c7b3f03fc0bad795262dd1e/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L91): A "Set<byte[]>" cannot contain a "byte[]" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2175)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L198](https://github.com/rsksmart/rskj/blob/a3f72ed52faf4ecc3c7b3f03fc0bad795262dd1e/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L198): Method org.ethereum.datasource.LevelDbDataSource.put(byte[], byte[]) passes a concatenated string to SLF4J's format string [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALO_APPENDED_STRING_IN_FORMAT_STRING)
1. ![CRITICAL][CRITICAL] [LevelDbDataSource.java#L203](https://github.com/rsksmart/rskj/blob/a3f72ed52faf4ecc3c7b3f03fc0bad795262dd1e/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L203): Method org.ethereum.datasource.LevelDbDataSource.put(byte[], byte[]) passes a concatenated string to SLF4J's format string [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3ALO_APPENDED_STRING_IN_FORMAT_STRING)
1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L89](https://github.com/rsksmart/rskj/blob/a3f72ed52faf4ecc3c7b3f03fc0bad795262dd1e/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L89): Method org.ethereum.datasource.DataSourceWithCache.addKeys(Set, KeyValueDataSource) uses array as basis of collection [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AABC_ARRAY_BASED_COLLECTIONS)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

rskops commented:
SonarQube analysis reported 2 issues

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L91](https://github.com/rsksmart/rskj/blob/21098faa858d9a31998457993f6871a4a824ce73/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L91): A "Set<byte[]>" cannot contain a "byte[]" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2175)
1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L89](https://github.com/rsksmart/rskj/blob/21098faa858d9a31998457993f6871a4a824ce73/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L89): Method org.ethereum.datasource.DataSourceWithCache.addKeys(Set, KeyValueDataSource) uses array as basis of collection [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AABC_ARRAY_BASED_COLLECTIONS)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

tinchou commented on rskj-core/src/main/resources/reference.conf:
```diff
@@ -299,4 +299,8 @@ prune {
     }
+}
+
+cache-size {
```
Consider hardcoding this number in code first, so we don't generate lots of config entries that need to be maintained.

diega replied:
we addressed this offline

tinchou commented on rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java:
```diff
@@ -182,15 +184,23 @@ public String getName() {
 
     @Override
     public byte[] put(byte[] key, byte[] value) {
+        if (key == null || value == null) {
```
`Objects.requireNonNull`

tinchou commented on rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java:
```diff
@@ -245,17 +255,29 @@ public void delete(byte[] key) {
+                byte[] key = entry.getKey().getData();
+                byte[] value =entry.getValue();
+                // Make this check explicit so that users know very clearly what to expect
+                if (Objects.isNull(key) || Objects.isNull(value)) {
```
Objects.requireNonNull

tinchou commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    public byte[] put(byte[] key, byte[] value) {
+        if (key == null || value == null) {
```
`Objects.requireNonNull`

SergioDemianLerner commented on rskj-core/src/main/java/co/rsk/config/RskSystemProperties.java:
```diff
@@ -417,4 +417,8 @@ public int getPeerP2PPingInterval(){
     }
+
+    public int getStatesCacheSize() {
+        return configFromFiles.getInt("cache-size.states");
```
Does this mean the config var is "states" in config section "cache-size" ? I seems that the logical thing is var "cache-size" in section "state". 
I does getInt() use a different scooping system ?

diega replied:
Yes, that's what it means. We discussed it briefly within the team and we decided to put it this way because it's more likely to have other caches configurations in the future than more _states_ tuning.
I'm not quite sure what do you mean by _scooping system_, would you mind to elaborate?

adauto82 commented on rskj-core/src/main/java/co/rsk/trie/TrieImpl.java:
```diff
@@ -1089,6 +1089,9 @@ private static boolean isEmptyTrie(byte[] value, TrieImpl[] nodes, Keccak256[] h
     public Trie getSnapshotTo(Keccak256 hash) {
         // This call shouldn't be needed since internally try can know it should store data
         //this.save();
+        if (getHash().equals(hash)) {
```
I disagree, is not totally unnecessary. The other way would be to make another method that does not copy this, and be very mindful of the use we give it. I would vote against it.

SergioDemianLerner replied:
FYI: This line provides an enormous speed boost for unitrie, if merged.

SergioDemianLerner commented on rskj-core/src/main/java/org/ethereum/datasource/HashMapDB.java:
```diff
@@ -72,21 +99,58 @@ public String getName() {
+            // Make this check explicit so that users know very clearly what to expect
+            Objects.requireNonNull(keyBytes);
+            Objects.requireNonNull(value);
+            if (keyBytes.length != 0) {
```
This is the wrong condition. The condition should be if (value.length != 0). This was copied from updateBatchInternal() which is wrong also (but never actually executed).
This bug was already reported (in a slack conversation). 
Red flag: There are no tests for this, being this critical to consensus.
Also there is no uniform and documented way to use updateBatch to delete items: if put(x,null) IS accepted, then updateBatch with null values should also. I'm in favor of accepting null and empty as synonyms (I know some core developers don't). In any case, coherence is key.

SergioDemianLerner commented on rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java:
```diff
@@ -245,27 +252,38 @@ public void delete(byte[] key) {
+                // Make this check explicit so that users know very clearly what to expect
+                Objects.requireNonNull(key);
+                Objects.requireNonNull(value);
+                if (key.length != 0) {
```
Buggy condition ? Shouldn't it be value.length !=0 ?

rskops commented:
SonarQube analysis reported 1 issue

1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L83](https://github.com/rsksmart/rskj/blob/c8fa4c206f90203d8cfb06246e2868b1f23b817b/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L83): Null passed for non-null parameter of HashMapDB.put(byte[], byte[]) in org.ethereum.datasource.DataSourceWithCache.delete(byte[]) [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF_ALL_TARGETS_DANGEROUS)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

rskops commented:
SonarQube analysis reported 1 issue

1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L84](https://github.com/rsksmart/rskj/blob/3b9b8f5585417f19de146db990f7b3a26e37d75c/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L84): Null passed for non-null parameter of HashMapDB.put(byte[], byte[]) in org.ethereum.datasource.DataSourceWithCache.delete(byte[]) [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=findbugs%3ANP_NULL_PARAM_DEREF_ALL_TARGETS_DANGEROUS)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

rskops commented:
SonarQube analysis reported 3 issues

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L52](https://github.com/rsksmart/rskj/blob/5aa9964577ba5a9ca8c01d8a00d3a7c7ce747514/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L52): This call to "contains()" may be a performance hot spot if the collection is large. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2250)
1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L88](https://github.com/rsksmart/rskj/blob/5aa9964577ba5a9ca8c01d8a00d3a7c7ce747514/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L88): This call to "remove()" may be a performance hot spot if the collection is large. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2250)
1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L53](https://github.com/rsksmart/rskj/blob/5aa9964577ba5a9ca8c01d8a00d3a7c7ce747514/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L53): Return an empty array instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

rskops commented:
SonarQube analysis reported 1 issue

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L26](https://github.com/rsksmart/rskj/blob/c43933ec3e5a9f367cd12c2bd44b4d863e02b9f5/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L26): Unused import - java.util.function.Predicate. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

SergioDemianLerner commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,149 @@
+
+        value = base.get(key);
+        if (value == null) {
+            committedCacheMap.put(ByteUtil.wrap(key), null);
```
This "if" is now redundant.

tinchou commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,146 @@
+    // commitedCacheHits = 248
+    // Processing 134 blocks grows the commitedCache to 100K entries, or approximately
+    // or 25 Mbytes. A cache of about 100 Mbytes seems rasonable. Anyway, for
+    // precaution we'll set the limit to 100K entries.
```
This comment is outdated

diega replied:
It's not outdated, it's just in the wrong place. I'll move it to the proper place so it gets sense in the right context. ty!

tinchou commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,146 @@
+
+    @Override
+    public byte[] get(byte[] key) {
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
```
`Objects.requireNonNull(key);`

tinchou commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,146 @@
+        Stream<ByteArrayWrapper> uncommittedKeys = uncommittedCacheMap.entrySet().stream()
+                .map(Map.Entry::getKey);
+
+        return Stream.concat(Stream.concat(baseKeys, committedKeys), uncommittedKeys)
```
[BUG] You shouldn't map and then collect because `byte[]`s doesn't implement deep equality

diega replied:
Great catch, thanks!

tinchou commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,146 @@
+                .filter(e -> e.getValue() == null)
+                .map(Map.Entry::getKey);
+        // there is no need to filter null values on uncommittedCacheMap
+        Stream<ByteArrayWrapper> uncommittedKeys = uncommittedCacheMap.entrySet().stream()
```
use `keySet()`

rskops commented:
SonarQube analysis reported 1 issue

1. ![MAJOR][MAJOR] [DetailsDataStore.java#L149](https://github.com/rsksmart/rskj/blob/b93ccd714b74240e854a9884bc718d33d01887e2/rskj-core/src/main/java/org/ethereum/db/DetailsDataStore.java#L149): Method org.ethereum.db.DetailsDataStore.flushInternal() does not presize the allocation of a collection [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APSC_PRESIZE_COLLECTIONS)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

diega commented:
@SergioDemianLerner I've addressed all your comments, thanks for your review. I've decided to change the `KeyValueDataSource#updateBatch` to receive two collections, one to update and one to remove. I also added a test for `DataSourceWithCache#delete` for catching one of your other comments

rskops commented:
SonarQube analysis reported 2 issues

1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/8a522dd47ff74dee190f5e912bfbea0bd991ecab/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/8a522dd47ff74dee190f5e912bfbea0bd991ecab/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

rskops commented:
SonarQube analysis reported 2 issues

1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/7fc0889c1df0cb0fd5d9e49c1cf5e03178cec93c/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/7fc0889c1df0cb0fd5d9e49c1cf5e03178cec93c/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

rskops commented:
SonarQube analysis reported 4 issues

1. ![CRITICAL][CRITICAL] [DataSourceWithCache.java#L26](https://github.com/rsksmart/rskj/blob/d4bc5a0cfdae70e8ef7942969546e89c27af86f4/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L26): Unused import - java.util.function.Predicate. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.imports.UnusedImportsCheck)
1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L49](https://github.com/rsksmart/rskj/blob/d4bc5a0cfdae70e8ef7942969546e89c27af86f4/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L49): Return an empty array instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/d4bc5a0cfdae70e8ef7942969546e89c27af86f4/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/d4bc5a0cfdae70e8ef7942969546e89c27af86f4/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

rskops commented:
SonarQube analysis reported 3 issues

1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L50](https://github.com/rsksmart/rskj/blob/8e7cb285a9cd2c7c3f91d3f350cb540879d1872c/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L50): Return an empty array instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/8e7cb285a9cd2c7c3f91d3f350cb540879d1872c/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/8e7cb285a9cd2c7c3f91d3f350cb540879d1872c/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

rskops commented:
SonarQube analysis reported 3 issues

1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L48](https://github.com/rsksmart/rskj/blob/df5744a5dfd861fb4afe9f5f3e7254caa72c038b/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L48): Return an empty array instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/df5744a5dfd861fb4afe9f5f3e7254caa72c038b/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/df5744a5dfd861fb4afe9f5f3e7254caa72c038b/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

rskops commented:
SonarQube analysis reported 3 issues

1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L48](https://github.com/rsksmart/rskj/blob/4a8e29126fc9b4c4ed305f6d4f061ebb23f4b582/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L48): Return an empty array instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/4a8e29126fc9b4c4ed305f6d4f061ebb23f4b582/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/4a8e29126fc9b4c4ed305f6d4f061ebb23f4b582/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

SergioDemianLerner commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,154 @@
+        }
+
+        byte[] value = base.get(key);
+        committedCache.put(wrappedKey, value);
```
null value, as expected, is allowed here to be stored in commitedCache

diega replied:
added as a comment

SergioDemianLerner commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,154 @@
+        }
+
+        // don't move unchanged keys unnecessarily
+        Map<ByteArrayWrapper, byte[]> updatedKeys = new HashMap<>(rows);
```
minor: updatedKeys sounds like a set. should be renamed something like: rowsToUpdate or entriesToUpdate

diega replied:
sure, renamed

rskops commented:
SonarQube analysis reported 3 issues

1. ![MAJOR][MAJOR] [DataSourceWithCache.java#L48](https://github.com/rsksmart/rskj/blob/3a5210495dd85566725534948227bf4d8b70a238/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L48): Return an empty array instead of null. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1168)
1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/3a5210495dd85566725534948227bf4d8b70a238/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/3a5210495dd85566725534948227bf4d8b70a238/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

SergioDemianLerner commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,155 @@
+        Objects.requireNonNull(key);
+        ByteArrayWrapper wrappedKey = ByteUtil.wrap(key);
+        // fully delete this element
+        committedCache.remove(wrappedKey);
```
Minor minor: prevention of deleted deletes

diega replied:
do you mean to add a `committedCache.containsKey(wrappedKey)` check before invoking the removal itself? I don't see what would be the benefit

SergioDemianLerner replied:
it prevents the  uncommittedCache.put(wrappedKey, null) to perform a later base.delete(), because we assume the base DB is slower than a committedCache.containsKey()

diega commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,155 @@
+
+        // don't move unchanged keys unnecessarily
+        Map<ByteArrayWrapper, byte[]> rowsToUpdate = new HashMap<>(rows);
+        rowsToUpdate.entrySet().removeIf(e -> Arrays.equals(e.getValue(), committedCache.get(e.getKey())));
```
> if a large batch of keys are removed several times, updateBatch won't detect it, and every time the uncommittedKeysToRemove will be loaded by the same keys
I cannot see how it's less optimal to implement a mechanism to detect recurrent removal attempts than just add that collection to the `uncommittedKeysToRemove` set. It may be more readable to make the uncommitedCache support null semantics tough. I'll implement this as a separate commit so you can review it, and if it looks better I'll squash that solution before merge

SergioDemianLerner replied:
Because the "base" is assumed to be slower (for deletes, even if the element does not exists), and that's why the cache is needed.

rskops commented:
SonarQube analysis reported 3 issues

1. ![MINOR][MINOR] [DataSourceWithCache.java#L118](https://github.com/rsksmart/rskj/blob/e925b70b48191374c210ec5a0426d2ee35864b51/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L118): Remove the parentheses around the "keyToRemove" parameter (sonar.java.source not set. Assuming 8 or greater.) [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1611)
1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/e925b70b48191374c210ec5a0426d2ee35864b51/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/e925b70b48191374c210ec5a0426d2ee35864b51/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

SergioDemianLerner commented on rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java:
```diff
@@ -0,0 +1,148 @@
+        Map<ByteArrayWrapper, byte[]> rowsToUpdate = new HashMap<>(rows);
+        rowsToUpdate.entrySet().removeIf(e -> Arrays.equals(e.getValue(), committedCache.get(e.getKey())));
+
+        committedCache.keySet().removeAll(rowsToUpdate.keySet());
```
Rephrasing the previous comment with pseudocode: here the code that follows should be:
**keysToCommitRemove = keysToRemove  excluding commitCache.keys(filter = value == null).**
committedCache.keySet().removeAll(keysToRemove);
uncommittedCache.putAll(rowsToUpdate);
**keysToCommitRemove**.forEach((keyToRemove) -> 
 uncommittedCache.put(keyToRemove, null));
to get optimum cache strategy.
But all this key filtering would not be needed if updateBatch() received a single list, with null values for deletions, as was in the original.

SergioDemianLerner replied:
Philosophic side note:  There should be a balance between the simplicity of an interface vs the simplicity of an implementation. If a simple interface leads to a complex implementation, then perhaps it's better to have a no-so-simple interface, with straightforward implementation.

SergioDemianLerner replied:
Final suggested code:
public void updateBatch(Map<ByteArrayWrapper, byte[]> rows, Set<ByteArrayWrapper> keysToRemove) {
         if (rows.containsKey(null) || rows.containsValue(null)) {
             throw new IllegalArgumentException("Cannot update null values");
  Map<ByteArrayWrapper, byte[]> combinedRows= new HashMap<>(rows);
 keysToRemove.forEach((keyToRemove) -> combinedRows.put(keyToRemove, 
 null));
 combinedRows.entrySet().removeIf(e -> Arrays.equals(e.getValue(), 
  uncommittedCache.get(e.getKey())));
 committedCache.removeAll(combinedRows);
 uncommitedCache.putAll(combinedRows);
}

rskops commented:
SonarQube analysis reported 3 issues

1. ![MINOR][MINOR] [DataSourceWithCache.java#L118](https://github.com/rsksmart/rskj/blob/1ff74728907f539044b4289f835ee0216e65c0e7/rskj-core/src/main/java/org/ethereum/datasource/DataSourceWithCache.java#L118): Remove the parentheses around the "keyToRemove" parameter (sonar.java.source not set. Assuming 8 or greater.) [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1611)
1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/1ff74728907f539044b4289f835ee0216e65c0e7/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/1ff74728907f539044b4289f835ee0216e65c0e7/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

rskops commented:
SonarQube analysis reported 2 issues

1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/7dcb78508202375fbbe644676d73c6f18a1971eb/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/7dcb78508202375fbbe644676d73c6f18a1971eb/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

rskops commented:
SonarQube analysis reported 2 issues

1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/ce585735334656ffb3c0204b1a6b3d983fe26ece/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/ce585735334656ffb3c0204b1a6b3d983fe26ece/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

rskops commented:
SonarQube analysis reported 2 issues

1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/a32c98d1ea364fbb6a767707daada527f6e1322f/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/a32c98d1ea364fbb6a767707daada527f6e1322f/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

rskops commented:
SonarQube analysis reported 4 issues

1. ![MAJOR][MAJOR] [TrieImpl.java#L814](https://github.com/rsksmart/rskj/blob/6646d4efed2962ca665d668f2ab32e607ade5037/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L814): Method co.rsk.trie.TrieImpl.split(TrieKeySlice) appears to call the same method on the same object redundantly [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MINOR][MINOR] [TrieImpl.java#L519](https://github.com/rsksmart/rskj/blob/6646d4efed2962ca665d668f2ab32e607ade5037/rskj-core/src/main/java/co/rsk/trie/TrieImpl.java#L519): Return a copy of "value". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2384)
1. ![INFO][INFO] [LevelDbDataSource.java#L281](https://github.com/rsksmart/rskj/blob/6646d4efed2962ca665d668f2ab32e607ade5037/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L281): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)
1. ![INFO][INFO] [LevelDbDataSource.java#L291](https://github.com/rsksmart/rskj/blob/6646d4efed2962ca665d668f2ab32e607ade5037/rskj-core/src/main/java/org/ethereum/datasource/LevelDbDataSource.java#L291): Constrained method org.ethereum.datasource.LevelDbDataSource.updateBatch(Map, Set) converts checked exception to unchecked [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3AEXS_EXCEPTION_SOFTENING_NO_CHECKED)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'
